### ISSUE TITLE
Ocaml versions older than 4.08 (included) do not build under GCC 10

### ISSUE THREAD

**Issue by tperami** [ocaml/ocaml#9622]
# Description 

Ocaml 4.08 (and older version) do not compile with default GCC 10.1.0 

I tried some other versions (`4.02.3`, `4.06.1` and `4.07.1`) via OPAM but not directly with their respective branches. 

Ocaml 4.09 builds fine both directly and with OPAM. I didn't try 4.10 because it's my system ocaml

# Steps to reproduce
```
git clone git@github.com:ocaml/ocaml.git
git checkout 4.08
./configure
make world.opt
```

# Expected behavior
 
It builds without errors.

# Actual behavior 

```
Lots of:
gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -Werror -fno-tree-vrp -g -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -DOCAML_STDLIB_DIR='"/usr/local/lib/ocaml"'  -o name.o name.c

Then:
rm -f libcamlrun_pic.a && ar rc libcamlrun_pic.a  interp_bpic.o misc_bpic.o stacks_bpic.o fix_code_bpic.o startup_aux_bpic.o startup_byt_bpic.o freelist_bpic.o major_gc_bpic.o minor_gc_bpic.o memory_bpic.o alloc_bpic.o roots_byt_bpic.o globroots_bpic.o fail_byt_bpic.o signals_bpic.o signals_byt_bpic.o printexc_bpic.o backtrace_byt_bpic.o backtrace_bpic.o compare_bpic.o ints_bpic.o floats_bpic.o str_bpic.o array_bpic.o io_bpic.o extern_bpic.o intern_bpic.o hash_bpic.o sys_bpic.o meta_bpic.o parsing_bpic.o gc_ctrl_bpic.o md5_bpic.o obj_bpic.o lexing_bpic.o callback_bpic.o debugger_bpic.o weak_bpic.o compact_bpic.o finalise_bpic.o custom_bpic.o dynlink_bpic.o spacetime_byt_bpic.o afl_bpic.o unix_bpic.o bigarray_bpic.o main_bpic.o && ranlib libcamlrun_pic.a
gcc -shared -o libcamlrun_shared.so interp_bpic.o misc_bpic.o stacks_bpic.o fix_code_bpic.o startup_aux_bpic.o startup_byt_bpic.o freelist_bpic.o major_gc_bpic.o minor_gc_bpic.o memory_bpic.o alloc_bpic.o roots_byt_bpic.o globroots_bpic.o fail_byt_bpic.o signals_bpic.o signals_byt_bpic.o printexc_bpic.o backtrace_byt_bpic.o backtrace_bpic.o compare_bpic.o ints_bpic.o floats_bpic.o str_bpic.o array_bpic.o io_bpic.o extern_bpic.o intern_bpic.o hash_bpic.o sys_bpic.o meta_bpic.o parsing_bpic.o gc_ctrl_bpic.o md5_bpic.o obj_bpic.o lexing_bpic.o callback_bpic.o debugger_bpic.o weak_bpic.o compact_bpic.o finalise_bpic.o custom_bpic.o dynlink_bpic.o spacetime_byt_bpic.o afl_bpic.o unix_bpic.o bigarray_bpic.o main_bpic.o -lm -ldl  -lpthread 
/usr/bin/ld: backtrace_bpic.o:/home/perami/Projects/REMS/ocaml/runtime/backtrace.c:31: multiple definition of `caml_debug_info'; backtrace_byt_bpic.o:/home/perami/Projects/REMS/ocaml/runtime/backtrace_byt.c:47: first defined here
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:310: libcamlrun_shared.so] Error 1
make[1]: Leaving directory '/home/perami/Projects/REMS/ocaml/runtime'
make: *** [Makefile:321: coldstart] Error 2
```

# Expected reason 

GCC 10 made the option `-fno-common` the default but `-fcommon` was the default on GCC 9 and lower. That means that, in GCC 10, C variables defined in headers without `extern` will end up doing multiple definitions.

# System specs

- GCC : `10.1.0`
- make: `4.3`
- Linux: `5.6.15`
- Distro: Archlinux updated on 2020-06-01
- System-wide ocaml: `4.10.0` (Should not influence this)
- OPAM: `2.0.7` (It fails with and without OPAM)

**Comment by tperami** [comment]
I tried to do 
```
./configure CFLAGS=-fcommon
```
But this seems to be fully ignored by the `configure` script and all the commands run by make do not contain this option.

**Comment by Octachron** [comment]
According to the discussion in https://discuss.ocaml.org/t/opam-switch-create-and-gcc10 , for 4.08
```ocaml
./configure ... CC="gcc -fcommon"
```
should work.
We will have to either fix the opam packages or patch the old branches with #9180. And it may make sense to release new bugfix versions.

**Comment by tperami** [comment]
@Octachron Thanks for the work around!
I didn't know that there was this website that I should have checked before posting the issue.
Maybe the issue should stay open until the problem is actually fixed? I'll let you decide

**Comment by Octachron** [comment]
It is a compiler issue that needs to be addressed, I am rather in favor of keeping it open.

**Comment by glondu** [comment]
This is related to https://github.com/ocaml/ocaml/issues/9540 . In Debian, I applied [this patch](https://salsa.debian.org/ocaml-team/ocaml/-/blob/debian/sid/debian/patches/0007-Avoid-duplicate-definitions-of-common-global-variabl.patch).

**Comment by SantiagoBautista** [comment]
I seem to have the same problem for ocaml 4.09.0 :
```
$ opam switch create 4.09.0 --jobs=1

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ocaml-base-compiler.4.09.0] found in cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] The compilation of ocaml-base-compiler failed at "/home/sbautist/.opam/opam-init/hooks/sandbox.sh build make -j1 world".
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base

#=== ERROR while compiling ocaml-base-compiler.4.09.0 =========================#
# context     2.0.6 | linux/x86_64 |  | https://opam.ocaml.org#ead61a76
# path        ~/.opam/4.09.0/.opam-switch/build/ocaml-base-compiler.4.09.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build make -j1 world
# exit-code   2
# env-file    ~/.opam/log/ocaml-base-compiler-46027-c61e7b.env
# output-file ~/.opam/log/ocaml-base-compiler-46027-c61e7b.out
### output ###
# [...]
# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp -g -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -DOCAML_STDLIB_DIR='"/home/sbautist/.opam/4.09.0/lib/ocaml"'  -o spacetime_byt_bpic.o spacetime_byt.c
# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp -g -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -DOCAML_STDLIB_DIR='"/home/sbautist/.opam/4.09.0/lib/ocaml"'  -o afl_bpic.o afl.c
# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp -g -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -DOCAML_STDLIB_DIR='"/home/sbautist/.opam/4.09.0/lib/ocaml"'  -o unix_bpic.o unix.c
# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp -g -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -DOCAML_STDLIB_DIR='"/home/sbautist/.opam/4.09.0/lib/ocaml"'  -o bigarray_bpic.o bigarray.c
# gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -fno-tree-vrp -g -fPIC -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -DOCAML_STDLIB_DIR='"/home/sbautist/.opam/4.09.0/lib/ocaml"'  -o main_bpic.o main.c
# rm -f libcamlrun_pic.a && ar rc libcamlrun_pic.a  interp_bpic.o misc_bpic.o stacks_bpic.o fix_code_bpic.o startup_aux_bpic.o startup_byt_bpic.o freelist_bpic.o major_gc_bpic.o minor_gc_bpic.o memory_bpic.o alloc_bpic.o roots_byt_bpic.o globroots_bpic.o fail_byt_bpic.o signals_bpic.o signals_byt_bpic.o printexc_bpic.o backtrace_byt_bpic.o backtrace_bpic.o compare_bpic.o ints_bpic.o floats_bpic[...]
# gcc -shared -o libcamlrun_shared.so interp_bpic.o misc_bpic.o stacks_bpic.o fix_code_bpic.o startup_aux_bpic.o startup_byt_bpic.o freelist_bpic.o major_gc_bpic.o minor_gc_bpic.o memory_bpic.o alloc_bpic.o roots_byt_bpic.o globroots_bpic.o fail_byt_bpic.o signals_bpic.o signals_byt_bpic.o printexc_bpic.o backtrace_byt_bpic.o backtrace_bpic.o compare_bpic.o ints_bpic.o floats_bpic.o str_bpic.o [...]
# /usr/bin/ld: backtrace_bpic.o:/home/sbautist/.opam/4.09.0/.opam-switch/build/ocaml-base-compiler.4.09.0/runtime/backtrace.c:31: multiple definition of `caml_debug_info'; backtrace_byt_bpic.o:/home/sbautist/.opam/4.09.0/.opam-switch/build/ocaml-base-compiler.4.09.0/runtime/backtrace_byt.c:47: first defined here
# collect2: error: ld returned 1 exit status
# make[1]: *** [Makefile:297: libcamlrun_shared.so] Error 1
# make[1]: Leaving directory '/home/sbautist/.opam/4.09.0/.opam-switch/build/ocaml-base-compiler.4.09.0/runtime'
# make: *** [Makefile:345: coldstart] Error 2
```

**Comment by lthls** [comment]
The discuss thread mentions that the first patched version is 4.09.1. However, apparently the opam repository only has patches for versions up to 4.08.1. If you're only trying to install a 4.09 version, go for 4.09.1 instead. (@avsm I assume the patch you applied to the 4.08 versions would also apply to 4.09.0)

**Comment by SantiagoBautista** [comment]
Thank you ! Version 4.09.1 does work for me.
